external_drivers/dhi_dfs/mdal_dhi_dfs.hpp

```diff
@@ -0,0 +1,369 @@
+    size_t m3DVertexCount = 0;
+    int mTotalVolumeCount = 0;
+    bool mIsDoublePrecision = false;
+    double mDoubleDeleteValue = 0.0;
+    float mFloatDeleteValue = 0.0;
+
+    //LevelValues mLevelValues;
+    std::vector<int> mVolumeCountPerTimeStep;
+    std::vector<int> mMaximumeLevelCountPerTimeStep;
+    std::vector<std::vector<int>> mFaceToStartVolumePositionPerTimeStep;
+    std::vector<std::vector<int>> mFaceLevelCountPerTimeStep;
+    std::vector<std::vector<double>> mFaceLevelsDataPerTimeStep;
+
+    std::vector<double> mRawDataDouble;
+    std::vector<float> mRawDataFloat;
+
+    void *rawDataPointerForRead( size_t size );
+    double rawDataValue( size_t i );
+    void buildVolumeForTimeStep( LONG timeStepNo );
+};
+
+
+
+class DatasetOnVolumes : public Dataset
+{
+  public:
+    DatasetOnVolumes( LPFILE Fp,
+                      LPHEAD  Pdfs,
+                      LONG timeStepNo,
+                      size_t maxSize,
+                      bool doublePrecision,
+                      double deleteDoubleValue,
+                      float deleteFloatValue,
+                      LevelValuesGenerator *levelValueGenerator )
+      : Dataset( Fp, Pdfs, timeStepNo, maxSize, doublePrecision, deleteDoubleValue, deleteFloatValue )
+      , mLevelValueGenerator( levelValueGenerator ) {}
+
+    virtual int volumeCount() override;
+    virtual int verticalLevelCountData( int indexStart, int count, int *buffer ) override;
+    virtual int verticalLevelData( int indexStart, int count, double *buffer ) override;
+    virtual int maximum3DLevelCount() override;
+    int faceToVolume( int indexStart, int count, int *buffer );
+    void unload() override;
+
+    template<typename T>
+    void reverseAndActiveData( std::vector<T> &dataArray, T deleteValue )
+    {
+      // need to reorder the data, MDAL need to have level data from top to bottom
+      const std::vector<int> &faceToStartVolumePosition = mLevelValueGenerator->faceTostartVolumePosition( mTimeStepNo );
+      const std::vector<int> &levelCounts = mLevelValueGenerator->levelCounts( mTimeStepNo );
+
+      assert( levelCounts.size() == faceToStartVolumePosition.size() );
+
+      mActive.resize( levelCounts.size() );
+
+      for ( size_t faceIndex = 0; faceIndex < faceToStartVolumePosition.size(); ++faceIndex )
+      {
+        size_t volumeStartIndex = faceToStartVolumePosition.at( faceIndex );
+        if ( levelCounts.at( faceIndex ) > 1 )
+        {
+          size_t volumeCount = levelCounts.at( faceIndex );
+          std::vector<T>::iterator it = dataArray.begin() + volumeStartIndex;
+          std::reverse( it, it + volumeCount );
+
+          for ( std::vector<T>::iterator itv = dataArray.begin() + volumeStartIndex; itv != dataArray.begin() + volumeStartIndex + volumeCount; ++itv )
+          {
+            bool isActive = ( *itv ) != deleteValue;
+            if ( !isActive )
+              ( *itv ) = 0.0;
```

Comment:
if it is not active, why we need to delete value? also can we extract NODATA val to some define, maybe even use -9999 as in other drivers?

Reply:
right, a left over, better to use` std::numeric_limits<double>::quiet_NaN()`